Integrate riscv-arch-test

xiaohan484 opened the pull request:
This work integrate riscv-arch-test with I,M,C,Zifencei,privilege to
rv32emu-next.Use `make compliance` to do compliance test with specific
extension test or test case.

Note: rv32emu-next has passed all test in I,M,Zifencei,
only c.ebreak fail in C and 9 tests fail in privilege.

Co-authored-by: YuJen-Lin <qscesz871107@gmail.com>

xiaohan484 committed:
Integrate riscv-arch-test

This work integrate riscv-arch-test with I,M,C,Zifencei,privilege to
rv32emu-next.Use `make compliance` to do compliance test with specific
extension test or test case.

Current progress of this emulator in riscv-arch-test(RV32):
* Passed Tests
    - `I`: Base Integer Instruction Set
    - `M`: Standard Extension for Integer Multiplication and Division
    - `Zifencei`: Instruction-Fetch Fence
* Failed Tests
    - `C`: Standard Extension for Compressed Instruction
        + `cebreak`
    - `privilege`: RISCV Privileged Specification
        + 2 system call
            * `ebreak`
            * `ecall`
        + 5 misaligned CB-type instructions
            * `misalign-beq`
            * `misalign-bge`
            * `misalign-bgeu`
            * `misalign-blt`
            * `misalign-bltu`
        + 1 misaligned CL-type instructions
            * `misalign-lw`
        + 1 misaligned CS-type instructions
            * `misalign-sw`
* Unsupported tests
    - `F` Standard Extension for Single-Precision Floating-Point

Co-authored-by: YuJen-Lin <qscesz871107@gmail.com>

jserv commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "riscv-arch-test"]
+	path = riscv-arch-test
+	url = https://github.com/riscv-non-isa/riscv-arch-test.git
```
Use `https://github.com/riscv-non-isa/riscv-arch-test` for short.

jserv commented on Makefile:
```diff
@@ -81,6 +81,21 @@ check: $(BIN)
 	(cd $(OUT); ../$(BIN) hello.elf)
 	(cd $(OUT); ../$(BIN) puzzle.elf)
 
+COMPLIANCE_DIR ?= ./riscv-arch-test 
```
Replace `COMPLIANCE_DIR` with `ARCH_TEST_DIR`.

jserv commented on Makefile:
```diff
@@ -81,6 +81,21 @@ check: $(BIN)
+export TARGETDIR = $(shell pwd)
+export XLEN = 32
+export JOBS ?= -j
+export WORK = $(TARGETDIR)/build/compliance
```
Ditto. Replace `compliance` with `arch-test`.

jserv commented on compliance-target/link.ld:
```diff
@@ -0,0 +1,16 @@
+  .bss : { *(.bss) }
+  _end = .;
+}
+
```
No trailing blank line.

jserv commented on compliance-target/model_test.h:
```diff
@@ -0,0 +1,27 @@
+
+#define RVMODEL_DATA_END \
+        .align 4; .global end_signature; end_signature:
+
```
Keep only one blank line.

jserv commented on main.c:
```diff
@@ -8,6 +8,10 @@
 static bool opt_trace = false;
 
+/* RISCV compliance test mode */
+static bool opt_compliance = false;
```
Ditto. Rename `opt_compliance` to `opt_arch_test`.

jserv commented on main.c:
```diff
@@ -101,7 +105,8 @@ static void print_usage(const char *filename)
             "Options:\n"
-            "  --trace : print executable trace\n",
+            "  --trace : print executable trace\n"
+            "  --compliance [filename] : dump signature to the given file, required by compliance test\n",
```
The option name should be `--arch-test`.

jserv commented on Makefile:
```diff
@@ -81,6 +81,21 @@ check: $(BIN)
 
+COMPLIANCE_DIR ?= ./riscv-arch-test 
+export RISCV_TARGET = compliance-target
+export RISCV_PREFIX = riscv-none-embed-
```
Rewrite as `export RISCV_PREFIX ?= riscv-none-embed-` which allows overriding.

jserv commented on Makefile:
```diff
@@ -81,6 +81,21 @@ check: $(BIN)
+	git submodule update --init
+
+compliance: $(BIN) $(COMPLIANCE_DIR)
+	$(Q)$(MAKE) --quiet -C $(COMPLIANCE_DIR) clean;
```
No `;` needed.

jserv commented on README.md:
```diff
@@ -30,6 +30,22 @@ make demo
 
+## Run riscv-arch-test
+
+Once the submodule `riscv-arch-test` is pulled, run all the available compliance test via command:
```
Improve the descriptions. You should mention the rationale and general RISC-V Arch Test motivation.

jserv commented on README.md:
````diff
@@ -30,6 +30,22 @@ make demo
+```shell
+make compliance RISCV_DEVICE=M RISCV_TEST=div-01
+```
+
````
Show the current coverage summary for reference.

xiaohan484 committed:
Revise riscv-arch-test description and Makefile

jserv commented on README.md:
```diff
@@ -30,6 +30,34 @@ make demo
 
+## riscv-arch-test
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) is basic set of tests that ensure software implemented can be match important aspects of the RISC-V specification.(not a **substitute for rigorous design verification**)
```
Check the grammar: "ensure software implemented can be match important aspects of ..."

jserv commented on README.md:
```diff
@@ -30,6 +30,31 @@ make demo
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) is basic set of tests that ensure software implemented can be match important aspects of the RISC-V specification.(not a **substitute for rigorous design verification**)
+
+There has the **reference signatures** generate by [RISC-V SAIL](https://github.com/riscv/sail-riscv) and the test signatures, specific memory location can be written.Once the test be executed, the **test signatures** will be written by testing-model(rv32emu-next)during the test, and it will be compared to **reference signature**.The test will pass if both signatures exactly match.
```
Ditto. Check the grammar.

jserv reviewed (changes requested):
Improve the writing!

jserv commented on README.md:
````diff
@@ -30,6 +30,32 @@ make demo
+```shell
+make arch-test RISCV_DEVICE=M RISCV_TEST=div-01
+```
+The detail in setup enviroment variables is in [RISC-V Architectural Testing Framework](https://github.com/riscv-non-isa/riscv-arch-test/blob/master/doc/README.adoc) **5.1 Setup environment variables**.
````
Should be "details"

jserv commented on arch-test-target/device/rv32i_m/I/Makefile.include:
```diff
@@ -0,0 +1,22 @@
+    -nostdlib \
+    -nostartfiles
+
+COMPILE_TARGET=\
```
Rewrite as `COMPILE_TARGET = \`

Be aware of the spaces.

jserv commented on arch-test-target/device/rv32i_m/M/Makefile.include:
```diff
@@ -0,0 +1,22 @@
+    -nostdlib \
+    -nostartfiles
+
+COMPILE_TARGET=\
```
Ditto.

jserv commented on arch-test-target/device/rv32i_m/Zifencei/Makefile.include:
```diff
@@ -0,0 +1,22 @@
+    -nostdlib \
+    -nostartfiles
+
+COMPILE_TARGET=\
```
Ditto.

jserv commented on arch-test-target/device/rv32i_m/privilege/Makefile.include:
```diff
@@ -0,0 +1,22 @@
+    -nostdlib \
+    -nostartfiles
+
+COMPILE_TARGET=\
```
Ditto.

jserv commented on main.c:
```diff
@@ -118,6 +123,16 @@ static bool parse_args(int argc, char **args)
                 opt_trace = true;
                 continue;
             }
+            if (!strcmp(arg, "--compliance")) {
```
Change the option name!

jserv commented on main.c:
```diff
@@ -177,6 +220,11 @@ int main(int argc, char **args)
     }
 
+    /* dump test result in test mode */
+    if (opt_compliance) {
```
Ditto.

jserv reviewed (changes requested):
The co-author should be `YuJen-Lin <qscesz871107@gmail.com>`

xiaohan484 committed:
Revise README.md and variable name

Revise the grammar of the riscv-arch-test description and replace the
"compliance" to "arch-test".

jserv commented on README.md:
````diff
@@ -30,6 +30,32 @@ make demo
+```shell
+make arch-test RISCV_DEVICE=M RISCV_TEST=div-01
+```
+The details in setup enviroment variables is in [RISC-V Architectural Testing Framework](https://github.com/riscv-non-isa/riscv-arch-test/blob/master/doc/README.adoc) **5.1 Setup environment variables**.
````
Check the grammar.

jserv replied:
You should make a list of known issues for RISC-V architecture tests.

jserv commented on README.md:
```diff
@@ -30,6 +30,32 @@ make demo
 
+## riscv-arch-test
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) the basic set of tests that can ensure the riscv model's behaviour that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
```
Add comma (`,`) after `riscv-arch-test`. Also, should add "is" before "the basic set of .."

jserv commented on README.md:
```diff
@@ -30,6 +30,32 @@ make demo
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test) the basic set of tests that can ensure the riscv model's behaviour that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
+
+There have **reference signatures** that is generated by the formal RISCV model [RISC-V SAIL](https://github.com/riscv/sail-riscv) and the test files *.elf. The .elf files contain several testing instructions, data, and **signature**.The **test signature** is the specific data location that need to be written during the test. Once the test is executed, the **test signatures** will be written by t ...
```
Ditto. Check the grammar.

jserv commented on main.c:
```diff
@@ -129,6 +144,34 @@ static bool parse_args(int argc, char **args)
+    }
+
+    /* use the entire .data section as a fallback */
+    elf_get_data_section_range(elf, &start, &end);
```
Append a new blank line after `elf_get_data_section_range(elf, &start, &end);`

xiaohan484 committed:
Revise the grammar of riscv-arch-test description

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+        + 2 system call
+            * `ebreak`
+            * `ecall`
+        + 5 misaligned CB-type
```
append "instructions" following "CB-type"

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+    - `C`: Standard Extension for Compressed Instruction
+        + `cebreak`
+    - `privilege`: RISCV Privileged Specification
+        + 2 system call
```
should be "system calls"

jserv reviewed (changes requested):
Fix wording

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
 
+## riscv-arch-test
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test), is the basic set of tests that can ensure the riscv model's behaviour that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
```
Let's use American English. That is, use "behavior" rather than "behaviour."

See https://grammarist.com/spelling/behavior-behaviour/

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test), is the basic set of tests that can ensure the riscv model's behaviour that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
+
+There have **reference signatures** that are generated by the formal RISCV model [RISC-V SAIL](https://github.com/riscv/sail-riscv) and the test files *.elf. The .elf files contain several testing instructions, data, and **signatures**. The **test signatures** are the specific data location that need to be written by testing model(rv32emu-next) during the test. Once the test is executed, the **te ...
```
Don't say "*.elf"

Instead, introduce "suffix" to rewrite. See https://en.wikipedia.org/wiki/Filename_extension

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+            * `misalign-bgeu`
+            * `misalign-blt`
+            * `misalign-bltu`
+        + 1 misaligned CL-type
```
Ditto. Append "instructions"

jserv commented:
> Note: rv32emu-next has passed all test in I,M,Zifencei, only c.ebreak fail in C and 9 tests fail in privilege.

Improve the descriptions about test results for current ISA coverage.

xiaohan484 committed:
Fix wording of README.md

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test), is the basic set of tests that can ensure the riscv model's behavior that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
+
+There have **reference signatures** that are generated by the formal RISCV model [RISC-V SAIL](https://github.com/riscv/sail-riscv) and the Executable and Linkable Format(`ELF`) files. The `ELF` files such as `cadd-01.elf` that contain several testing instructions, data, and **signatures**. The **test signatures** are the specific data location that need to be written by testing model(rv32emu-nex ...
```
Change `RISCV` to `RISC-V`.

jserv replied:
Remove the word "rv32emu-next" since this project will be renamed soon.

jserv replied:
Instead, you can say "this emulator" or "this project."

xiaohan484 committed:
Rename "rv32emu-next" and fix wording

jserv commented on elf.c:
```diff
@@ -249,6 +269,15 @@ const char *elf_find_symbol(elf_t *e, uint32_t addr)
     return c_map_at_end(e->symbols, &it) ? NULL : c_map_iter_value(&it, char *);
 }
 
+bool elf_get_data_section_range(elf_t *e, uint32_t *start, uint32_t *end) {                                
```
Indent with `clang-format -i`.

jserv commented on README.md:
```diff
@@ -30,6 +30,57 @@ make demo
+
+The RISC-V Architectural Tests, [riscv-arch-test](https://github.com/riscv-non-isa/riscv-arch-test), is the basic set of tests that can ensure the risc-v model's behavior that matches RISC-V specifications while implementing specific software.(not a **substitute for rigorous design verification**)
+
+There have **reference signatures** that are generated by the formal RISC-V model [RISC-V SAIL](https://github.com/riscv/sail-riscv) and the Executable and Linkable Format(`ELF`) files. The `ELF` files such as `cadd-01.elf` that contain several testing instructions, data, and **signatures**. The **test signatures** are the specific data location that need to be written by testing model(this emula ...
```
Change "There have **reference signatures** that are generated " to
> There are **reference signatures** generated

jserv replied:
Change "data location" to "data locations"

xiaohan484 committed:
Indent clang format and revise README.md

jserv commented:
Thank @xiaohan484 and @sammer1107 for contributing!
